pkg/broker/instance_operations.go

```diff
@@ -188,9 +193,104 @@ func (b *Broker) createOrUpdateResources(ctx context.Context, client *mongodbatl
 		}
 	}
 
+	for _, endpoint := range dp.PrivateEndpoints {
+		if endpoint.ID != "" {
+			// endpoint service already created
+			continue
+		}
+
+		logger.Debugw("start creation process for PrivateEndpoints", "project ID", p.ID)
+		conn, _, err := client.PrivateEndpoints.Create(ctx, p.ID, &mongodbatlas.PrivateEndpointConnection{
+			ProviderName: endpoint.Provider,
+			Region:       endpoint.Region,
+		})
+		if err != nil {
+			return errors.Wrap(err, "cannot create Private Endpoint Service")
+		}
+
+		endpoint.ID = conn.ID
+	}
+
 	return nil
 }
 
+// TODO: this retry logic is clunky, come up with something better?
+func (b *Broker) postCreateResources(ctx context.Context, client *mongodbatlas.Client, dp *dynamicplans.Plan) (retry bool, err error) {
+	logger := b.funcLogger()
+
+	logger.Debugw("Setup PrivateEndpoints", "PrivateEndpoints", dp.PrivateEndpoints)
+	for _, endpoint := range dp.PrivateEndpoints {
+		atlasService, _, err := client.PrivateEndpoints.Get(ctx, dp.Project.ID, endpoint.Provider, endpoint.ID)
+		if err != nil {
+			return false, errors.Wrap(err, "cannot add Private Endpoint to Endpoint Service")
+		}
+
+		switch atlasService.Status {
+		case "INITIATING", "WAITING_FOR_USER":
+			retry = true
+
+			continue
+
+		case "AVAILABLE":
+			break
+
+		default:
+			return false, errors.Wrapf(err, "Private Endpoint service is in the wrong state: %s", atlasService.Status)
+		}
+
+		logger.Debugw("Creating private endpoint", "endpoint", endpoint.ID)
+		future, err := privateendpoint.Create(ctx, endpoint, atlasService)
+		if err != nil {
+			return false, errors.Wrap(err, "cannot create Private Endpoint in Azure")
+		}
+
+		pe, err := future()
+		if err != nil {
+			logger.Debugw("PrivateEndpoint not ready; retry", "PrivateEndpoint", pe)
+			retry = true
+
+			continue
+		}
+
+		addr, err := privateendpoint.GetIPAddress(ctx, pe, endpoint)
+		if err != nil {
+			return false, errors.Wrap(err, "cannot get IP address for Private Endpoint")
+		}
+
+		existing, r, err := client.PrivateEndpoints.GetOnePrivateEndpoint(ctx, dp.Project.ID, endpoint.Provider, endpoint.ID, *pe.ID)
+		if err != nil {
+			if r == nil || r.StatusCode != http.StatusNotFound {
+				return false, errors.Wrap(err, "cannot get Private Endpoint")
+			}
+		}
+
+		if existing == nil {
+			_, _, err = client.PrivateEndpoints.AddOnePrivateEndpoint(ctx, dp.Project.ID, endpoint.Provider, endpoint.ID, &mongodbatlas.InterfaceEndpointConnection{
+				ID:                       *pe.ID,
+				PrivateEndpointIPAddress: addr,
+			})
+			if err != nil {
+				return false, errors.Wrap(err, "cannot add Private Endpoint to Endpoint Service")
+			}
+
+			continue
+		}
+
+		if existing.PrivateEndpointIPAddress == addr {
+			continue
+		}
+
+		_, err = client.PrivateEndpoints.DeleteOnePrivateEndpoint(ctx, dp.Project.ID, endpoint.Provider, endpoint.ID, *pe.ID)
```

Comment:
[q] why are we deleting the private endpoint?

Reply:
Not quite sure, but my best guess is that it prevents half-configured endpoints from interrupting the next try.

If everything goes fine we return/continue before this line. If something isn't correct we clear the endpoint before returning with `retry == true`.

Reply:
ok, this makes sense 👍  - can we add the comment please?